src/fillalgebra.jl

```diff
@@ -211,61 +211,47 @@ function dot(u::AbstractVector{T}, D::Diagonal{U,<:Zeros}, v::AbstractVector{V})
+        promote_shape(a,b)
+        return elconvert(promote_op(+,T,V),a)
+    end
+    @eval +(a::Zeros, b::$TYPE) = b + a
 end
--(a::Zeros, b::Zeros) = -(a + b)
--(a::Ones, b::Ones) = Zeros(a)+Zeros(b)
 
-# Zeros +/- Fill and Fill +/- Zeros
-function +(a::AbstractFill{T}, b::Zeros{V}) where {T, V}
-    size(a) ≠ size(b) && throw(DimensionMismatch("dimensions must match."))
-    return AbstractFill{promote_type(T, V)}(a)
-end
-+(a::Zeros, b::AbstractFill) = b + a
--(a::AbstractFill, b::Zeros) = a + b
--(a::Zeros, b::AbstractFill) = a + (-b)
-
-# Zeros +/- Array and Array +/- Zeros
-function +(a::Zeros{T, N}, b::AbstractArray{V, N}) where {T, V, N}
-    size(a) ≠ size(b) && throw(DimensionMismatch("dimensions must match."))
-    return AbstractArray{promote_type(T,V),N}(b)
-end
-function +(a::Array{T, N}, b::Zeros{V, N}) where {T, V, N}
-    size(a) ≠ size(b) && throw(DimensionMismatch("dimensions must match."))
-    return AbstractArray{promote_type(T,V),N}(a)
+# for VERSION other than 1.6, could use ZerosMatrix only
+function +(a::AbstractFillMatrix{T}, b::UniformScaling) where {T}
+    n = checksquare(a)
+    return a + Diagonal(Fill(zero(T) + b.λ, n))
 end
 
-function -(a::Zeros{T, N}, b::AbstractArray{V, N}) where {T, V, N}
-    size(a) ≠ size(b) && throw(DimensionMismatch("dimensions must match."))
-    return -b + a
-end
--(a::Array{T, N}, b::Zeros{V, N}) where {T, V, N} = a + b
+# LinearAlgebra defines `-(a::AbstractMatrix, b::UniformScaling) = a + (-b)`,
+# so the implementation of `-(a::UniformScaling, b::AbstractFill{<:Any,2})` is sufficient
+-(a::UniformScaling, b::AbstractFill) = -b + a # @test I-Zeros(3,3) === Diagonal(Ones(3))
 
+-(a::Ones, b::Ones) = Zeros(a) + Zeros(b)
 
-+(a::AbstractRange, b::Zeros) = b + a
+# necessary for AbstractRange, Diagonal, etc
++(a::AbstractFill, b::AbstractFill) = fill_add(a, b)
++(a::AbstractFill, b::AbstractArray) = fill_add(b, a)
++(a::AbstractArray, b::AbstractFill) = fill_add(a, b)
+-(a::AbstractFill, b::AbstractFill) = a + (-b)
+-(a::AbstractFill, b::AbstractArray) = a + (-b)
+-(a::AbstractArray, b::AbstractFill) = a + (-b)
 
-function +(a::ZerosVector{T}, b::AbstractRange) where {T}
-    size(a) ≠ size(b) && throw(DimensionMismatch("dimensions must match."))
-    Tout = promote_type(T, eltype(b))
-    return Tout(first(b)):Tout(step(b)):Tout(last(b))
-end
-function +(a::ZerosVector{T}, b::UnitRange) where {T<:Integer}
-    size(a) ≠ size(b) && throw(DimensionMismatch("dimensions must match."))
-    Tout = promote_type(T, eltype(b))
-    return AbstractUnitRange{Tout}(b)
+@inline function fill_add(a, b::AbstractFill)
+    promote_shape(a, b)
+    a .+ getindex_value(b)
 end
 
-function -(a::ZerosVector, b::AbstractRange)
-    size(a) ≠ size(b) && throw(DimensionMismatch("dimensions must match."))
-    return -b + a
-end
--(a::AbstractRange, b::ZerosVector) = a + b
+@inline elconvert(::Type{T}, A::AbstractRange) where T = T(first(A)):T(step(A)):T(last(A))
```

Comment:
Add comment that this is a work around for `convert(AbstractArray{Float64}, 1:5)` not returning a range.

Reply:
I just noticed that we already have `_copy_oftype` defined that does something similar to what `elconvert` does. It'll be good to merge the two.

Reply:
will have a look.

Reply:
> I just noticed that we already have `_copy_oftype` defined that does something similar to what `elconvert` does. It'll be good to merge the two.

It doesn't work since
```julia
julia> _copy_oftype(1:5,Float64)
5-element Vector{Float64}:
 1.0
 2.0
 3.0
 4.0
 5.0
```